### ISSUE TITLE
eth_estimateGas requires gas.

### ISSUE THREAD

**Issue by MicahZoltu** [ConsenSys-archive/ganache-cli-archive#411]
## Expected Behavior
When calling `eth_estimateGas`, no gas should be required.

## Current Behavior
Calling `eth_estimateGas` without providing a `gas` property on the transaction object will result in a failure if the transaction requires more than 90,000 gas.

## Possible Solution
Based on the behavior of both Geth and Parity, local call executions (meaning `eth_call` and `eth_estimateGas`) should always succeed even when no gas is provided.

## Steps to Reproduce (for bugs)
1. Start `TestRpc.server({ gasLimit: '0x632EA0' })`
2. Call `eth_estimateGas` via JSON-RPC over HTTP with a large contract upload:
```
{"id":4694917462827,"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"from":"0x496BA8A4aD3E2539e03a54aBBc7f8A5bBcD8dA69","data":"0x60606040526003805460ff19169055341561001957600080fd5b60008054600160a060020a03338116600160a060020a031990921691909117808355168152600160208190526040909120805460ff19169091179055610b63806100646000396000f300606060405236156101255763ffffffff7c01000000000000000000000000000000000000000000000000000000006000350416630a89e0aa811461012a5780630c6b87b11461015157806322763ae1146101765780633b048123146101895780633f08882f1461019c5780634e94c829146101bb57806351180c63146101ea578063575b35e0146102005780635f436da21461022b57806363a599a41461025057806375f12b21146102635780637ef502981461027657806386d1a69f1461028c5780638ab1d6811461029f5780638da5cb5b146102be5780639a2cf0cc146102d15780639b19251a14610338578063b9e290f914610357578063e43252d714610379578063f2fde38b14610398578063f39ec1f7146103b7578063ff55da1c146103cd575b600080fd5b341561013557600080fd5b61013d6103ef565b604051901515815260200160405180910390f35b341561015c57600080fd5b61013d600160a060020a0360043581169060243516610459565b341561018157600080fd5b61013d610530565b341561019457600080fd5b61013d610549565b34156101a757600080fd5b61013d600160a060020a036004351661055d565b34156101c657600080fd5b6101ce61058c565b604051600160a060020a03909116815260200160405180910390f35b34156101f557600080fd5b61013d6004356105bc565b341561020b57600080fd5b61013d600160a060020a03600435811690602435811690604435166105f7565b341561023657600080fd5b61013d600160a060020a03600435811690602435166106d7565b341561025b57600080fd5b61013d61078a565b341561026e57600080fd5b61013d6107ca565b341561028157600080fd5b6101ce6004356107d3565b341561029757600080fd5b61013d6107ee565b34156102aa57600080fd5b61013d600160a060020a036004351661082b565b34156102c957600080fd5b6101ce61085d565b34156102dc57600080fd5b61013d600160a060020a0360048035821691602480359091169190606490604435908101908301358060208181020160405190810160405280939291908181526020018383602002808284375094965061086c95505050505050565b341561034357600080fd5b61013d600160a060020a0360043516610971565b341561036257600080fd5b61013d600435600160a060020a0360243516610986565b341561038457600080fd5b61013d600160a060020a03600435166109d1565b34156103a357600080fd5b61013d600160a060020a0360043516610a05565b34156103c257600080fd5b6101ce600435610aaf565b34156103d857600080fd5b61013d600160a060020a0360043516602435610aca565b6000805433600160a060020a0390811691161461040b57600080fd5b60008054600160a060020a031681526001602052604090205460ff16151561043257600080fd5b5060008054600160a060020a03168152600160208190526040909120805460ff1916905590565b6000805433600160a060020a0390811691161461047557600080fd5b60008054600160a060020a031681526001602052604090205460ff16151561049c57600080fd5b82600160a060020a031663551cacfd836000604051602001526040517c010000000000000000000000000000000000000000000000000000000063ffffffff8416028152600160a060020a039091166004820152602401602060405180830381600087803b151561050c57600080fd5b6102c65a03f1151561051d57600080fd5b5050506040518051506001949350505050565b60035460009060ff161561054357600080fd5b50600190565b60035460009060ff16151561054357600080fd5b600160a060020a03811660009081526001602052604081205460ff16151561058457600080fd5b506001919050565b60006105b77f4175677572000000000000000000000000000000000000000000000000000000610aaf565b905090565b60006105c73361055d565b50506000908152600260205260409020805473ffffffffffffffffffffffffffffffffffffffff19169055600190565b6000805433600160a060020a0390811691161461061357600080fd5b60008054600160a060020a031681526001602052604090205460ff16151561063a57600080fd5b83600160a060020a031663ed6b2d7d84846000604051602001526040517c010000000000000000000000000000000000000000000000000000000063ffffffff8516028152600160a060020a03928316600482015291166024820152604401602060405180830381600087803b15156106b257600080fd5b6102c65a03f115156106c357600080fd5b505050604051805150600195945050505050565b6000805433600160a060020a039081169116146106f357600080fd5b60008054600160a060020a031681526001602052604090205460ff16151561071a57600080fd5b82600160a060020a03166392eefe9b836000604051602001526040517c010000000000000000000000000000000000000000000000000000000063ffffffff8416028152600160a060020a039091166004820152602401602060405180830381600087803b151561050c57600080fd5b6000805433600160a060020a039081169116146107a657600080fd5b60035460ff16156107b657600080fd5b506003805460ff1916600190811790915590565b60035460ff1681565b600260205260009081526040902054600160a060020a031681565b6000805433600160a060020a0390811691161461080a57600080fd5b60035460ff16151561081b57600080fd5b506003805460ff19169055600190565b60006108363361055d565b5050600160a060020a03166000908152600160208190526040909120805460ff1916905590565b600054600160a060020a031681565b6000805433600160a060020a0390811691161461088857600080fd5b60008054600160a060020a031681526001602052604090205460ff1615156108af57600080fd5b83600160a060020a0316630d10d43c84846000604051602001526040518363ffffffff167c01000000000000000000000000000000000000000000000000000000000281526004018083600160a060020a0316600160a060020a0316815260200180602001828103825283818151815260200191508051906020019060200280838360005b8381101561094c578082015183820152602001610934565b505050509050019350505050602060405180830381600087803b15156106b257600080fd5b60016020526000908152604090205460ff1681565b60006109913361055d565b505060008281526002602052604090208054600160a060020a03831673ffffffffffffffffffffffffffffffffffffffff19909116179055600192915050565b60006109dc3361055d565b5050600160a060020a03166000908152600160208190526040909120805460ff19168217905590565b6000805433600160a060020a03908116911614610a2157600080fd5b60008054600160a060020a031681526001602052604090205460ff1615610a7d5760008054600160a060020a039081168252600160208190526040808420805460ff199081169091559286168452909220805490911690911790555b5060008054600160a060020a03831673ffffffffffffffffffffffffffffffffffffffff199091161790556001919050565b600090815260026020526040902054600160a060020a031690565b600081815260026020526040812054600160a060020a0384811691161480610b23575060005433600160a060020a039081169116148015610b23575060008054600160a060020a031681526001602052604090205460ff165b1515610b2e57600080fd5b506001929150505600a165627a7a72305820c1ff7a10775f1266b5877486bcd3e9d6f9c9af605f40683359977e4491bf10880029"}]}
```
3. Notice that you receive a failure response:
```
{"message":"Error: base fee exceeds gas limit ...","code":-32000}
```

## Context
I am unable to run my integration tests or test my deployment scripts against TestRPC because I am using eth_estimateGas to figure out how much gas the functions need before I submit them to the chain.

## Your Environment
* Version used: 6.0.1
* Environment name and version (e.g. PHP 5.4 on nginx 1.9.1): node v8.5.0
* Operating System and version: Windows 10
* Link to your project: https://github.com/AugurProject/augur-core

**Comment by benjamincburns** [comment]
Hi Micah,

I agree that this would be nice. I find that needing to specify gas is quite annoying. However, since we're a test tool we rather pedantically follow the JSON RPC specification, even if some clients like geth and parity might not.

As the wiki says, [`eth_call`](https://github.com/ethereum/wiki/wiki/JSON-RPC#parameters-24) and [`eth_estimateGas`](https://github.com/ethereum/wiki/wiki/JSON-RPC#parameters-25) both have the optional `gas` parameter. Since it's a little unclear in the spec as to how this parameter is meant to be used in this case, and since the implementation of the EVM we use tracks and limits execution based on gas, even for `eth_call` and `eth_estimateGas`, we elect to treat it the same as `eth_sendTransaction`. Similarly, if the `gas` isn't specified, we use the same default gas limit (90,000) which we'd use for `eth_sendTransaction`.

The question to me in this case isn't whether we behave the same as geth or Parity, but rather it's whether code written to run against TestRPC will run successfully against these or other clients. If geth, parity, or other clients error when you specify `gas` on `eth_call` or `eth_estimateGas`, those clients likely have a bug.

Does this make sense? Feel free to ping me on Gitter if you want to discuss out of band.

**Comment by MicahZoltu** [comment]
**TL;DR**: You shouldn't use the wiki as a spec, and if you _are_ using the wiki as a spec you are not following it.

The wiki for `eth_estimateGas` says:
> If no gas limit is specified geth uses the block gas limit from the pending block as an upper bound.

This does not align with the behavior of TestRPC currently, which uses 90,000 as its limit.

As far as using the Wiki as a spec, I generally recommend against that since as you have indicated it is not clear enough to actually use as a spec for implementing a client and people trying to use it as a spec has lead to many cross-client incompatibilities in the past (e.g., between Geth and Parity and MetaMask).  I urge you to not try to use the Wiki as a spec and instead either do what you believe is reasonable _or_ do what the other clients do.

In this case, I do not think that defaulting to 90k gas is _reasonable_ for `eth_estimateGas`.  I think that Geth's behavior of defaulting to block gas limit _is_ somewhat reasonable as it makes the call finite (can't loop forever) while also being as useful as possible.  Personally, I would do 1.5x pending block gas limit as that would allow you to estimate gas for transactions that are borderline as to whether they will fit into a block or not.  This can be useful during development when you are trying to get a transaction to fit under the gas limit, being able to know how much gas it currently needs is valuable so you know how close you are).

**Comment by benjamincburns** [comment]
Your suggestions do sound reasonable, and I'm happy with the change you suggest. It'll certainly make things a bit more convenient.

When I read the text you quoted, I read it as using what geth does as an example for the purpose of illustration rather than limitation/prescription. I think it's just as reasonable to read it as a prescriptive example, however.

I think your advice on regarding the wiki as a spec is good in the short term, but what you're really highlighting is the need for a more well-defined spec from the Ethereum Foundation. In the absence of a spec, and in the absence of existing convergent implementation amongst the other clients, I don't think that doing what's reasonable will necessarily converge into something interoperable as well as we'd all like. I think it amounts to guessing what the other clients will decide upon.

If the foundation won't do it, I think the major clients need to get together to do it. I'm happy to help campaign for that.

**Comment by MicahZoltu** [comment]
> I think your advice on regarding the wiki as a spec is good in the short term, but what you're really highlighting is the need for a more well-defined spec from the Ethereum Foundation.

Yeah, I'm in 100% agreement with this.  Having a real spec for the JSON-RPC would help a _lot_ IMO, and give a mechanism for getting new features added to the RPC and give implementers something to rally around when it comes time to debate "which implementation is right and which is wrong".  I can't offer to dedicate any significant time to such a project, but I will advocate for it if you do start something up.  😄

**Comment by tcoulter** [comment]
From the sidelines: 

👍 to all this convo.

>  I can't offer to dedicate any significant time to such a project, but I will advocate for it if you do start something up
 
@benjamincburns I'm 💯 on you devoting time to this. 👌

**Comment by benjamincburns** [comment]
Thanks, @tcoulter — let’s talk later in the week on how best to go about it. I definitely want to work with the foundation as much as possible, and I have some ideas that I think you’ll like.

**Comment by benjamincburns** [comment]
Fixed by trufflesuite/ganache-core#25.

Thanks for sticking with this one, @MicahZoltu!
